src/centrality.rs

```diff
@@ -0,0 +1,217 @@
+        let p_w = predecessors.get_mut(&w).unwrap();
+        for v in p_w {
+            let iv = (*v).index();
+            delta[iv] += (sigma[&iv] as f64) * coeff;
+        }
+        if iw != is {
+            betweenness[iw] += delta[iw];
+        }
+    }
+}
+
+fn _accumulate_endpoints(
+    betweenness: &mut Vec<f64>,
+    node_count: usize,
+    sorted_by_distance: &Vec<NodeIndex>,
+    mut predecessors: HashMap<NodeIndex, Vec<NodeIndex>>,
+    sigma: HashMap<usize, i64>,
+    is: usize,
+) {
+    betweenness[is] += (sorted_by_distance.len() - 1) as f64;
+    let mut delta = vec![0.0; node_count];
+    for w in sorted_by_distance {
+        let iw = (*w).index();
+        let coeff = (1.0 + delta[iw]) / (sigma[&iw] as f64);
+        let p_w = predecessors.get_mut(&w).unwrap();
+        for v in p_w {
+            let iv = (*v).index();
+            delta[iv] += (sigma[&iv] as f64) * coeff;
+        }
+        if iw != is {
+            betweenness[iw] += delta[iw] + 1.0;
+        }
+    }
+}
+
+fn shortest_path_for_centrality<G>(
+    graph: G,
+    node_s: &G::NodeId,
+) -> (
+    Vec<NodeIndex>,
+    HashMap<G::NodeId, Vec<G::NodeId>>,
+    HashMap<usize, i64>,
+)
+where
+    G: NodeIndexable
+        + IntoNodeIdentifiers
+        + IntoNeighborsDirected
+        + GraphBase<NodeId = NodeIndex>, // for get() and get_mut()
+{
+    let mut sorted_by_distance: Vec<NodeIndex> = Vec::new(); // a stack
+    #[allow(non_snake_case)]
+    let mut Q: Vec<NodeIndex> = Vec::new(); // a queue
+    let mut predecessors = HashMap::<G::NodeId, Vec<G::NodeId>>::new();
+    let mut sigma = HashMap::<usize, i64>::new();
+    let mut distance = HashMap::<G::NodeId, i64>::new();
+
+    let i_s = graph.to_index(*node_s);
+    let index_s = NodeIndex::new(i_s);
+
+    for node in graph.node_identifiers() {
+        let i = graph.to_index(node);
+        let index = NodeIndex::new(i);
+        predecessors.insert(index, Vec::new());
+        sigma.insert(i, 0);
+        distance.insert(index, -1);
+    }
+    sigma.insert(index_s.index(), 1);
+    distance.insert(index_s, 0);
+    Q.push(index_s);
+    while !Q.is_empty() {
+        let v = Q.remove(0);
+        sorted_by_distance.push(v);
+        //        let neighbors = graph.neighbors_directed(v, Incoming);
+        for w in graph.neighbors(v) {
+            if *(distance.get(&w).unwrap()) < 0 {
+                Q.push(w);
+                distance.insert(w, *(distance.get(&v).unwrap()) + 1);
+            }
+            if *(distance.get(&w).unwrap()) == *(distance.get(&v).unwrap()) + 1
+            {
+                sigma.insert(
+                    graph.to_index(w),
+                    *(sigma.get(&graph.to_index(w)).unwrap())
+                        + *(sigma.get(&graph.to_index(v)).unwrap()),
+                );
+                let e_p = predecessors.get_mut(&w).unwrap();
+                e_p.push(v);
+            }
+        }
+    }
+    return (sorted_by_distance, predecessors, sigma);
```

Comment:
You don't actually need the return here, the last line in a function without a semicolon is an implicit return (`cargo clippy` will complain about this).

```suggestion
    (sorted_by_distance, predecessors, sigma)
```

Reply:
(lost this comment earlier?) I am usually in the always-use-explicit-return camp for languages that allow both. One reason is I sometimes see a big blob of computation at the bottom of a function and it's not clear whether its for the side effect or a return value. A strong convention helps clarify it. But, it's more important to follow conventions.

Reply:
Yeah I tend to agree with you I like to be explicit instead of implicit behavior, but I was deferring to the default linter rules that `cargo clippy` has and using an implicit return where possible seems to be the normal rust convention.